File: raft/grpc_client.go
package raft

import (
	"bytes"
	"context"
	"time"

	"github.com/eraft-io/eraft/labgob"
	"github.com/eraft-io/eraft/raftpb"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

type RaftgRPCClient struct {
	client raftpb.RaftServiceClient
	conn   *grpc.ClientConn
}

func NewRaftgRPCClient(addr string) *RaftgRPCClient {
	conn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		panic(err)
	}
	client := raftpb.NewRaftServiceClient(conn)
	return &RaftgRPCClient{client: client, conn: conn}
}

func (c *RaftgRPCClient) RequestVote(args *RequestVoteRequest, reply *RequestVoteResponse) bool {
	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
	defer cancel()
	req := &raftpb.RequestVoteRequest{
		Term:         int64(args.Term),
		CandidateId:  int64(args.CandidateId),
		LastLogIndex: int64(args.LastLogIndex),
		LastLogTerm:  int64(args.LastLogTerm),
	}
	resp, err := c.client.RequestVote(ctx, req)
	if err != nil {
		return false
	}
	reply.Term = int(resp.Term)
	reply.VoteGranted = resp.VoteGranted
	return true
}

func (c *RaftgRPCClient) AppendEntries(args *AppendEntriesRequest, reply *AppendEntriesResponse) bool {
	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
	defer cancel()
	entries := make([]*raftpb.Entry, len(args.Entries))
	for i, e := range args.Entries {
		var command []byte
		if e.Command != nil {
			var ok bool
			command, ok = e.Command.([]byte)
			if !ok {
				// Fallback: if it's not []byte, try to encode it using labgob
				// This handles cases where old code might have stored structs in the log
				w := new(bytes.Buffer)
				enc := labgob.NewEncoder(w)
				if err := enc.Encode(e.Command); err == nil {
					command = w.Bytes()
				}
			}
		}
		entries[i] = &raftpb.Entry{
			Index:   int64(e.Index),
			Term:    int64(e.Term),
			Command: command,
		}
	}
	req := &raftpb.AppendEntriesRequest{
		Term:         int64(args.Term),
		LeaderId:     int64(args.LeaderId),
		PrevLogIndex: int64(args.PrevLogIndex),
		PrevLogTerm:  int64(args.PrevLogTerm),
		LeaderCommit: int64(args.LeaderCommit),
		Entries:      entries,
	}
	resp, err := c.client.AppendEntries(ctx, req)
	if err != nil {
		return false
	}
	reply.Term = int(resp.Term)
	reply.Success = resp.Success
	reply.ConflictIndex = int(resp.ConflictIndex)
	reply.ConflictTerm = int(resp.ConflictTerm)
	return true
}

func (c *RaftgRPCClient) InstallSnapshot(args *InstallSnapshotRequest, reply *InstallSnapshotResponse) bool {
	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
	defer cancel()
	req := &raftpb.InstallSnapshotRequest{
		Term:              int64(args.Term),
		LeaderId:          int64(args.LeaderId),
		LastIncludedIndex: int64(args.LastIncludedIndex),
		LastIncludedTerm:  int64(args.LastIncludedTerm),
		Data:              args.Data,
	}
	resp, err := c.client.InstallSnapshot(ctx, req)
	if err != nil {
		return false
	}
	reply.Term = int(resp.Term)
	return true
}

File: raft_cpp/src/grpc_client.cpp
// grpc_client.cpp — gRPC client implementation
// Corresponds to Go: grpc_client.go

#include "raft/grpc_client.h"

#include <chrono>

namespace raft {

RaftgRPCClient::RaftgRPCClient(const std::string& addr)
    : channel_(grpc::CreateChannel(addr, grpc::InsecureChannelCredentials())),
      stub_(raftpb::RaftService::NewStub(channel_)) {}

RaftgRPCClient::~RaftgRPCClient() = default;

bool RaftgRPCClient::RequestVote(const RequestVoteRequest& args,
                                 RequestVoteResponse& reply) {
    grpc::ClientContext ctx;
    ctx.set_deadline(std::chrono::system_clock::now() +
                     std::chrono::milliseconds(100));

    raftpb::RequestVoteRequest req;
    req.set_term(args.term);
    req.set_candidate_id(args.candidate_id);
    req.set_last_log_index(args.last_log_index);
    req.set_last_log_term(args.last_log_term);

    raftpb::RequestVoteResponse resp;
    auto status = stub_->RequestVote(&ctx, req, &resp);
    if (!status.ok()) return false;

    reply.term        = static_cast<int>(resp.term());
    reply.vote_granted = resp.vote_granted();
    return true;
}

bool RaftgRPCClient::AppendEntries(const AppendEntriesRequest& args,
                                   AppendEntriesResponse& reply) {
    grpc::ClientContext ctx;
    ctx.set_deadline(std::chrono::system_clock::now() +
                     std::chrono::milliseconds(100));

    raftpb::AppendEntriesRequest req;
    req.set_term(args.term);
    req.set_leader_id(args.leader_id);
    req.set_prev_log_index(args.prev_log_index);
    req.set_prev_log_term(args.prev_log_term);
    req.set_leader_commit(args.leader_commit);

    for (const auto& e : args.entries) {
        auto* pe = req.add_entries();
        pe->set_index(e.index);
        pe->set_term(e.term);
        if (!e.command.empty()) {
            pe->set_command(e.command.data(), e.command.size());
        }
    }

    raftpb::AppendEntriesResponse resp;
    auto status = stub_->AppendEntries(&ctx, req, &resp);
    if (!status.ok()) return false;

    reply.term          = static_cast<int>(resp.term());
    reply.success       = resp.success();
    reply.conflict_index = static_cast<int>(resp.conflict_index());
    reply.conflict_term  = static_cast<int>(resp.conflict_term());
    return true;
}

bool RaftgRPCClient::InstallSnapshot(const InstallSnapshotRequest& args,
                                     InstallSnapshotResponse& reply) {
    grpc::ClientContext ctx;
    ctx.set_deadline(std::chrono::system_clock::now() +
                     std::chrono::seconds(1));

    raftpb::InstallSnapshotRequest req;
    req.set_term(args.term);
    req.set_leader_id(args.leader_id);
    req.set_last_included_index(args.last_included_index);
    req.set_last_included_term(args.last_included_term);
    if (!args.data.empty()) {
        req.set_data(args.data.data(), args.data.size());
    }

    raftpb::InstallSnapshotResponse resp;
    auto status = stub_->InstallSnapshot(&ctx, req, &resp);
    if (!status.ok()) return false;

    reply.term = static_cast<int>(resp.term());
    return true;
}

} // namespace raft
